src/components/TodosContext.tsx

```diff
@@ -0,0 +1,158 @@
+  updateTodoTitle: async () => {},
+  setFilter: () => {},
+  errorMessage: '',
+  setErrorMessage: () => {},
+  handlerDeleteCompleted: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  processingIds: [],
+  setProcessingIds: () => {},
+  inputRef: React.createRef<HTMLInputElement>(),
+});
+
+type Props = {
+  children: React.ReactNode,
+};
+
+export const TodoProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<Status>(Status.ALL);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [processingIds, setProcessingIds] = useState<number[]>([]);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+  // const [loading, setLoading] = useState<boolean>(false);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      });
+  }, []);
+
+  const addTodo = (todo: Todo) => {
+    setTodos([...todos, todo]);
+  };
+
+  const deleteTodo = (idTodo: number) => {
+    setProcessingIds((prevIds) => [...prevIds, idTodo]);
+
+    deleteTodoo(idTodo)
+      .then(() => {
+        setTodos((prevTodos) => prevTodos.filter((todo) => todo.id !== idTodo));
+      })
+      .catch(() => setErrorMessage('Unable to delete a todo'))
+      .finally(() => {
+        setProcessingIds([]);
+      });
+  };
+
+  const handlerDeleteCompleted = () => {
+    const completedTodoIds = todos
+      .filter(todo => todo.completed).map(todo => todo.id);
+
+    if (completedTodoIds.length === 0) {
+      return;
+    }
+
+    completedTodoIds.forEach(todoId => {
+      deleteTodoo(todoId)
+        .then(() => {
+          setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+        })
+        .catch(() => setErrorMessage('Unable to delete a todo'));
+    });
+  };
+
+  const toggleCompleted = (id: number) => {
+    setProcessingIds(prev => [...prev, id]);
+
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (!todoToUpdate) {
+      setProcessingIds([]);
+
+      return;
+    }
+
+    const updatedTodo = { ...todoToUpdate, completed: !todoToUpdate.completed };
+
+    updateTodo(updatedTodo)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.map(todo => (todo.id === id
+          ? { ...todo, completed: updatedTodo.completed }
+          : todo)));
+      })
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => {
+        setProcessingIds([]);
+      });
+  };
+
+  const toggleAllCompleted = () => {
+    if (activeTodos.length !== 0) {
+      activeTodos.forEach(todo => toggleCompleted(todo.id));
+    } else {
+      completedTodos.forEach(todo => toggleCompleted(todo.id));
+    }
+  };
+
+  const clearCompleted = () => {
```

Comment:
I think you don't use this function, please check

if its correct please delete it